packages/next/server/render.tsx

```diff
@@ -1096,7 +1093,6 @@ export async function renderToHTML(
     disableOptimizedLoading,
     head: documentResult.head,
     headTags: documentResult?.headTags,
-    styles: documentResult.styles,
```

Comment:
Unfortunately, I think we need to keep passing `styles` through to `getInitialProps` and to the `Document` props for backward compatibility. I'd work around the hook issue by doing:

```
const StyleRegistryCallbackContext = React.createContext(() => {})
const Container = ({ props }) => {
  const styleRegistry = useStyleRegistry()
  const callback = useContext(StyleRegistryCallbackContext)
  callback(styleRegistry)
  return children
}
```

And then for `AppTree`, make sure to wrap it in a no-op `StyleRegistryCallbackContext.Provider`. Wrap the other one in a provider that saves it to a local variable. Then read & return `styles` in `renderPage` (instead of `pages/_document.tsx`) via the local `styleRegistry` that got stored as a side effect.